File: q30-q39/q31/q31.cpp
#include <iostream>
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int i = nums.size() - 2;
        while (i >= 0 && nums[i+1] <= nums[i]){
            --i;
        }
        if (i >= 0){
            int j = nums.size() - 1;
            while (j >= 0 && nums[j] <= nums[i]){
                --j;
            }
            int temp = nums[j];
            nums[j] = nums[i];
            nums[i] = temp;
        }
        int j = nums.size() - 1;
        i ++;
        while( i < j ){
            int temp = nums[j];
            nums[j] = nums[i];
            nums[i] = temp;
            i++;
            j--;
        }
    }
};

File: q1-q99/q30-q39/q31/q31.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 2
        while i >= 0 and nums[i+1] <= nums[i]:
            i -= 1
        if i >= 0:
            j = len(nums) - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1
            temp = nums[j]
            nums[j] = nums[i]
            nums[i] = temp
        j = len(nums) - 1
        i += 1
        while i < j:
            temp = nums[j]
            nums[j] = nums[i]
            nums[i] = temp
            i += 1
            j -= 1
        
